sale_cutoff_time_delivery/models/sale_order.py

```diff
@@ -0,0 +1,123 @@
+                    "No cutoff applied on order %s as partner %s is set to use "
+                    "%s and warehouse %s doesn't apply cutoff." % (
+                        self.order_id,
+                        partner,
+                        partner.order_delivery_cutoff_preference,
+                        self.order_id.warehouse_id
+                    )
+                )
+            else:
+                _logger.warning(
+                    "No cutoff applied on order %s. %s time not applied"
+                    "on line %s." % (
+                        self.order_id, partner.order_delivery_cutoff_preference, self
+                    )
+                )
+            return res
+        tz = cutoff.get('tz')
+        date_planned = res.get("date_planned")
+        if tz and tz != 'UTC':
+            cutoff_time = time(hour=cutoff.get('hour'), minute=cutoff.get('minute'))
+            # Convert here to naive datetime in UTC
+            tz_loc = pytz.timezone(tz)
+            tz_date_planned = date_planned.astimezone(tz_loc)
+            tz_cutoff_datetime = datetime.combine(tz_date_planned, cutoff_time)
+            utc_cutoff_datetime = tz_utils.tz_to_utc_naive_datetime(tz_loc, tz_cutoff_datetime)
+        else:
+            utc_cutoff_datetime = date_planned.replace(
+                hour=cutoff.get('hour'), minute=cutoff.get('minute'), second=0
+            )
+        if date_planned <= utc_cutoff_datetime:
+            # Postpone delivery for date planned before cutoff to cutoff time
+            new_date_planned = date_planned.replace(
+                hour=utc_cutoff_datetime.hour, minute=utc_cutoff_datetime.minute, second=0
+            )
+        # Postpone delivery for order confirmed after cutoff to day after
+        else:
+            new_date_planned = date_planned.replace(
+                hour=utc_cutoff_datetime.hour, minute=utc_cutoff_datetime.minute, second=0
+            ) + timedelta(days=1)
+        _logger.debug(
+            "%s applied on order %s. Date planned for line %s"
+            " rescheduled from %s to %s" % (
+                partner.order_delivery_cutoff_preference, self.order_id,
+                self, date_planned, new_date_planned
+            )
+        )
+        res["date_planned"] = new_date_planned
+        return res
+
+    def _expected_date(self):
+        """Postpone expected_date to next cut-off"""
+        expected_date = super()._expected_date()
+        cutoff = self.order_id.get_cutoff_time()
+        if not cutoff:
+            return expected_date
+        cutoff_tz = cutoff.get('tz')
+        cutoff_time = time(hour=cutoff.get('hour'), minute=cutoff.get('minute'))
+        if cutoff_tz:
+            utc_cutoff_time = tz_utils.tz_to_utc_time(
+                cutoff_tz,
+                cutoff_time,
+                base_date=expected_date
+            )
+        else:
+            utc_cutoff_time = cutoff_time
+        if expected_date.time() <= utc_cutoff_time:
```

Comment:
You should not compare times. Apply here the same implementation you did in _prepare_procurement_values.
1:00 < 3:00 in CET so should remains the same after conversion to UTC: 23:00 <> 1:00

Reply:
I'm not sure to understand why this wouldn't work since `utc_cutoff_time` is computed according to `expected_date`. Can you please explain a bit more?

Reply:
@jbaudoux ping

Reply:
Let's define a cutoff at 6AM AWST (Australia time). You order at 10AM, so you are after the cutoff.
Now I convert to UTC: Cutoff=22PM UTC. Order time is 2AM. I am before the cutoff.